### ISSUE TITLE
[Feature Request] rclone mount: Cache the entire drive structure at startup, and add new files by monitoring Google's ChangeLog

### ISSUE THREAD

**Issue by jpat0000** [ncw/rclone#897]
[in reference to to this post on the rclone forum](https://forum.rclone.org/t/google-drive-quota-reached-seemingly-linked-to-folder-scanning/271) 

Excessively using queries seemingly results in exceeding Google Drive's quota, locking downloads for 24 hours. Caching the entire drive structure, and adding new files by periodically checking the Google Drive changelog should dramatically reduce the queries used. It would also significantly increase the speed which applications such as Plex, or Sonarr can check for new files by scanning every folder.

**Comment by yrh1** [comment]
I am using google drive with rclone as well and had same issue. Not too technical but understood that it can be done by keeping a copy of cache in memory. I will be ok with that as well.

**Comment by jkaberg** [comment]
Perhaps this could be tied in with https://github.com/ncw/rclone/issues/894 aswell?

**Comment by toomuchio** [comment]
Could be very useful for ACD as well, since if you hammer that too much on inbound requests you get a similar effect

**Comment by jrarseneau** [comment]
This would be incredibly useful for ACD as well, for like Plex, which scans. all. the. freaking. time and having these files/folders in a cache would improve the scanning speed immensely.

The cache can be updated when new files are written, modified, deleted, etc., i.e.: rclone copy/sync/move/delete/etc operations.

**Comment by pmow** [comment]
I second ACD as well.  A few weeks ago Amazon tells me I'm locked out due to "unusual activity" when I've never shared a file, and only have one machine mounting it.  Yesterday I noticed files not reading due to a 403 Forbidden error on GDrive, which turned out to be quota issue.  Also no files being shared.  This would be of immense help.

**Comment by martinbowling** [comment]
This would make rclone, google drive & plex so usable please please please 

Will donate to charity, give tips, whatever it might take

**Comment by IKShadow** [comment]
I support this as long as its optional, still have the bad taste with acd_cli and database corruptions :)

**Comment by scoopydude2002** [comment]
Which database type are you planning to use? Please consider SQLite

**Comment by jtl5770** [comment]
Just a question: for which remotes your part 2) can actually be implemented? Has ACD something comparable to get to changes happening to the file storage? Because just relying on monitoring the own copy/sync/whatever commands from the local rsync to know about changes would lead to changes from other places to the same remote not being visible....

**Comment by TechJennings** [comment]
this please like, more than anything in the universe this! ;)

**Comment by ncw** [comment]
@jtl5770 ACD has a [changes API](https://developer.amazon.com/public/apis/experience/cloud-drive/content/changes) which I haven't studied in detail, but it looks like it might do the job!

**Comment by felixbuenemann** [comment]
The python acd_cli uses a local node cache using a sqlite database which is updated by polling the changes api.

I think it would be great to also have this feature for normal rclone usage, not just FUSE, it would reduce API calls a lot if copy/sync is done continuously.

This feature would also open up the possibility for implementing a continuous sync feature in the future, which could sync from cloud to cloud or cloud to local using change streams and filesystem notifications for local to cloud.

The following remotes should be able to support change streams:

* Google Drive ([Changes v2](https://developers.google.com/drive/v2/reference/changes), [Changes v3](https://developers.google.com/drive/v3/reference/changes))
* Amazon Drive ([Changes API](https://developer.amazon.com/public/apis/experience/cloud-drive/content/changes))
* OneDrive ([Sync changes](https://dev.onedrive.com/items/view_delta.htm))
* Dropbox ([v1: /delta](https://www.dropbox.com/developers-v1/core/docs#delta), [v2: /list_folder](https://www.dropbox.com/developers/documentation/http/documentation#files-list_folder))

**Comment by ncw** [comment]
@felixbuenemann I agree 100% with you that a local DB fed by the changes is the way to go for rclone in general, not just mount.

I want to get rid of the concept of `root` out of the remotes first which will simplify them and will make this a lot easier .

I then imagine an optional interface to poll for changes on the remote which can be used to update a local database.  There are a few things that need to be thought out, like serializing objects

Thanks for looking up the APIs :-)

**Comment by jrarseneau** [comment]
Any chance of an ETA for this? 1 month, 3 months out?

Many of us are using node-gdrive-fuse or google-drive-ocamlfuse to mount Google as rclone will always result in a ban when. using Plex+rclone+Google Drive.

I think many of us are chomping at the bit to simplify.

Thanks @ncw !

**Comment by ncw** [comment]
I'd like to fit it into the next release which should be in a couple of months, but that might be over optimistic!

**Comment by spencercharest** [comment]
@ncw when you begin working on this feature would it be possible to prioritize Google Drive and push out a beta since Drive has the most issues caused by excessive API hits?

**Comment by felixbuenemann** [comment]
Btw. those Plex users might give Plex Cloud a try, it's still a bit buggy but generally works. Of course that will only work for Plex Pass subscribers with unencrypted data in Google Drive.

**Comment by Eleatmelon** [comment]
@felixbuenemann So, using Plex Cloud won't cause problem like this? As it is also performs scans of Google Drive.

**Comment by felixbuenemann** [comment]
It seems to work fine with my medium sized library (~1500 Movies). I'm also not using subdirectories per movie, which reduces API calls for listings a lot. Note that library updates on Plex Cloud are super sluggish, either to avoid rate limits or because of limited CPU/RAM.

**Comment by jpat0000** [comment]
@Eleatmelon I have read that people have gotten the 24h ban using Plex Cloud as well after refreshing their large Plex libraries.

@felixbuenemann Moving your movies into a single directory is responsible for why you haven't gotten the 24h ban yet. That's 1,500 less requests with each scan.

**Comment by FlipABit** [comment]
The problem here isn't specifically related to number of API requests in general. It's to do with Plex's media analyzation, not the directory scanning. When Plex analyzes your media files, it actually looks at several chunks of each file to gather information like resolution, codec, etc. On a regular hard drive, that's no big deal, but for network mounted drives in apps like rclone or NetDrive, that translates to a dozen or more HTTP Range requests, each asking GDrive for a small chunk of file. Even though only small chunks are being downloaded for each file, Google considers each range request as a "download" of the file. So when you have a large library with an initial media scan and analyzation, Google sees you "downloading" your 5-10GB+ media files dozens of times each in a very short period. This triggers their abuse prevention limit (different from their standard API rate limiting) and they purposefully don't publish what exactly that limit is (and if you talk to their support, they will say they can't tell you what it is). You can see examples of this happening by doing an initial scan of a Plex library on a mounted GSuite drive. If you have access to the GSuite admin console, you can go to "Reports" -> Audit (in the left hand menu) -> Drive. During the scan you'll see each media file listed multiple times and in the column next to it will say "<your name> downloaded an item". Plex Cloud did run into this when they first introduced GDrive support. They mitigated the issue by preventing media analyzation during the initial file scan, and only analyzing a file when you individually select an item to view in a Plex client. While the feature request in this thread could be useful, it will not likely solve the actual problem. You have a better shot by requesting that the Plex devs add the same option to the standard Plex Media Server that they did to their Plex Cloud server to limit media analyzation to be only on-demand.

**Comment by jpat0000** [comment]
@FlipABit I get what you're saying, and common sense would dictate that Google would have more a problem with accessing lots of files instead of too many API requests. However, my experience (which you can read on the rclone forum, it's linked at the top) suggests that this is strongly related to accessing a lot of folders, not media analyzing/accessing many files. 

tl;dr: 

A: I have used node-gdrive-fuse which operates similarly to rclone with the exception that it caches the file structure. No bans.

B: I have analyzed 25TBs worth of media, twice, in a single day using node-gdrive-fuse. No bans.

C: I have used over 10TB of Google Drive bandwidth in a single day. No bans.

D: Using rclone, I attempt to analyze all of my library, I get banned part way through.

D: Using rclone I have scanned my library for new content a few times in a day, while no media analyzing is necessary as all metadata had just painfully been generated. 24h ban. Let the ban ware off, test rclone again, no new media, just scanning the ol' library a few times, 24h ban. 

I am curious if Plex has stated that they changed the way media analyzation is done for the sake of avoiding 24 hour bans, or you're simply using that as anecdotal evidence. Of course, there are plenty of other reasons for Plex to make that change, such as conserving bandwidth, instead of needlessly wasting it on media which may never be watched.

Moreover, there are people who are saying that they are getting the 24h ban using Plex Cloud with Google Drive. Feel free to give it a google search and let me know if I am missing anything.

**Comment by Eleatmelon** [comment]
I did some test with my G Suite account. I think the problem is about 'Download', as Google said when you're banned: Download quota exceeded... Basically, it is what FlipABit said.

Each time Plex scan a new TV show (Only new added or if you changed your directory, the old files won't be scanned by doing update or refresh jobs even if you unmount Google Drive & re-mount it), every single media file will be labeled 'donwloaded' for about at least 3 times, some even 5~10 times. That is to say if you add a new library (1000 episodes for example), Google will consider you've downloaded 5000 large files in a short period. Then you get banned.

But not only Plex scanning will trigger this multiple download issue, playback of the media files also triggers multiple download -- even worse. I've tried to play a movie using Plex, and within 10 minutes, Google shows me that I've already downloaded it 20 times. I assume Plex does not download the files at a time, apparently that is how things work, it download a small section of the media files as buffer each time, and Google sees it as download of the files. I also test it with some other app such as nPlayer on iOS, it performs a little bit better, I watched a show and Google only showed 3~5 times of downloading that file. However, since you don't watch 100 or 200 episodes a day, I think it is just safe(to be proved) to store your media data on Google Drive and watch it through streaming. But having this kind of restriction will make Google Drive not a perfect choice of building a cloud media library.

**Comment by FlipABit** [comment]
> But not only Plex scanning will trigger this multiple download issue, playback of the media files also triggers multiple download -- even worse. I've tried to play a movie using Plex, and within 10 minutes, Google shows me that I've already downloaded it 20 times. I assume Plex does not download the files at a time, apparently that is how things work, it download a small section of the media files as buffer each time, and Google sees it as download of the files. 

This is another thing I have observed as well and is reproducible. Specifically, if you Direct Play a file via the Plex web player in Safari or an Apple TV 4, it will constantly access pieces of the file. A 2-hour Direct Played movie will stop approximately half way through when the ban kicks in. This does _not_ seem to occur when Direct Playing the same file in Chrome or a Fire TV Plex client. This is something Plex Cloud has not resolved.

There may very well be multiple types of bans here that we're all lumping into a single cause. Caching the file directory may help with one type of ban, but I guarantee people are still going to run into the "download quota exceeded" issue. And as to @jpat0000 "C" point, what I've described has nothing to do with the actual amount of data downloaded. It's the fact that Google _thinks_ large files are being downloaded multiple times each within seconds of each other, as @Eleatmelon has also observed and you can see an example of here: http://forums.plex.tv/discussion/comment/1379563/#Comment_1379563

> I am curious if Plex has stated that they changed the way media analyzation is done for the sake of avoiding 24 hour bans, or you're simply using that as anecdotal evidence

First, yes, they did explicitly say this: http://forums.plex.tv/discussion/comment/1383107/#Comment_1383107 and was listed in the December update (#177): http://forums.plex.tv/discussion/comment/1317102/#Comment_1317102. I was one of the first people to run into the issue with Plex Cloud (but not the only) and worked directly with the devs and Google support on the issue. The Plex devs were able to reproduce it themselves as well. While there are a small handful of people still experiencing a ban, Plex is unable to reproduce it after putting in the fix and none of those people have been able to provide logs showing that Plex Cloud is the culprit. Some have even admitted they're _also_ using rclone and/or rsync with a mounted GDrive elsewhere, which would explain some of it.

Like I said though, there may be a couple causes of the ban happening here, but the media analysis is definitely one of them.

**Comment by felixbuenemann** [comment]
FYI in my case google drive bans where usually caused by excessive data transfer when mirroring from google drive to amazon drive with rclone and not from api requests. I checked by using my own google drive api keys and requests were way below the limits. I haven't yet done any testing on how much data transferred over what time period exactly causes bans, so that I can avoid it with `--bwlimit`.

**Comment by pabru** [comment]
I'm also having a problem related to this.

My ACD account isn't even that full, but it absolutely murders Nautilus on Ubuntu 16.04. I suspected this may be the case if rclone does metadata fetch on each access, and it seems that's correct unfortunately. 

I've tried acd_cli to compare, and the local db makes a lot of difference - I can actually browse the FUSE mount easily without delays. As of right now, the only smooth-feeling ways to browse is via the command line and Dolphin (dolphin4) apparently.

**Comment by AloiSama** [comment]
I have the same problem with locks and then I checked my Google Admin reports feature. It showed each movie/episode being downloaded many many times during library updates (which doesn't agree with my server's download statistics). Could plex be asking for part of the file multiple times and google sees it as a whole file download? If so what would the best solution be?

I should also mention I'm using my own API credentials which reports my requests are below the limit.

Thanks,
Ali

**Comment by ncw** [comment]
I think #1346 will also help with this issue for google drive

**Comment by Wydarzenia** [comment]
I'm not using RClone for media, but still the lack of cache is a drawback. NCW, is there any agenda when/if it will be added?

**Comment by snickers2k** [comment]
+1 .. most important feature-request
i have no chance to use my GDrive the way it should be. Filebot and Plex constantly killing the service with too many requests. Plexmount is really an odd workaround (not working for me with encrypted drive).

**Comment by breunigs** [comment]
Polling Google Drive for changes while mounted has been implemented a while back already (a2e3af0523d6a4c3f01b8ddaa1b43939a863ecf7) and it is on by default, but can be configured trough `--poll-interval <Duration>`. To get the most out of it, you'll need to set a high `--dir-cache-time` of course.

The 2nd part of this ticket asks for all directories to be read preemptively. I'm not sure how that would be helping to not get banned, since it will need to query each directory. Rclone already implements a better approach if the remotes support it, but Drive does not apparently. I would suggest to see if this is really needed, since Plex/Sonarr seem to be scanning everything anyway?

The poll changes made it into 1.37 already, so can you guys give it a spin and report how it works for you?

**Comment by leonelwilliams** [comment]
So I could set `dir-cache-time` to something crazy like `30y` but it wouldn't negatively affect anything since `poll-interval` would keep everything up to date?

**Comment by breunigs** [comment]
That's the idea.

**Comment by gryphonmyers** [comment]
Can anyone confirm that the changes from a2e3af0 prevent the 24 hour ban when used with Plex?

**Comment by leonelwilliams** [comment]
Plex's scanner reads (downloads) a bit of each uncatalogued media file to determine things like resolution, language, subtitles, etc. A large `dir-cache-time` won't help with that and you'll still get a ban it seems. It could help if the library's imported and all Plex is doing is reading the directory listings looking for new content.

If you have a big library it looks like the only thing you can do is scan, get banned, wait 24 hours, repeat until the majority of your media has been catalogued. Plex Cloud seems to have a direct integration with Google Drive that doesn't trigger their bans. Unfortunately Plex hasn't released the connector/plugin they're using for Plex Cloud to ordinary PMS users.

**Comment by RikuXan** [comment]
I think you are right on the money and this is the biggest puzzle piece in preventing those GDrive bans. 
 
Plex scanner probing a media file results in about 11 download calls for me (according to the GDrive admin console), but I cannot imagine that much data actually flows (should be only headers plus a few hundred frames at most), the requests are probably only for small data ranges within the file.

The question then becomes whether or not the 11 download requests count towards the quota with their full size each time, only the actually transferred data or something inbetween (e.g. size of a chunk).  
If they count fully, abuse prevention kicking in is well understandable. For a 3GB file the scanner probably takes 10 seconds and would add 33GB to the quota in that time, a whopping 26GBit/s far outside normal usage territory.  
If they calculate the quota correctly, I can't imagine the ban reason to be volume-based since normal streaming should create at least the same load, however it might also react to the amount of download requests, irrespective of the transferred data.

If the guys over at Plex don't have a special agreement with Google that allows them to circumvent the abuse prevention mechanisms, it would indeed be very interesting how they managed to trim their system to not trigger abuse prevention.

**Comment by Trentflix** [comment]
I read somewhere (can't find source) that they just do a light-dummy scan initially to get the media in, but don't do the full scan (like get bit-rate, codec information etc.) until you actually click on the item. This is how they avoid the trigger.

**Comment by pmow** [comment]
What about prefetching the specific blocks ahead of time as a background process, that plex will need?  That way when they're needed they can be read from the cache.

**Comment by RikuXan** [comment]
Something like that could help. I'd argue though, as this is an application-specific cache, that such a feature would be better suited for something like Plexdrive rather than rclone, which is more of a generic solution.

**Comment by ncw** [comment]
I'm please to announce that after a lot of hard work by @remusb we have merged the first version of a cache backend for rclone.

Please have a go and check out [the docs](https://github.com/ncw/rclone/blob/master/docs/content/cache.md) and have a go!

Please report any problems in a [new issue on github](https://github.com/ncw/rclone/issues/new).

You'll find support in [the latest beta](https://beta.rclone.org).

I'm currently working on proper read/write support for mount which will also involve some caching - this isn't quite merged yet though.

**Comment by felixbuenemann** [comment]
This is looking great!

I've read through the docs and the only thing I'm missing so far is the ability to limit the amount of disk storage that gets used for caching, assuming your cache location is much smaller than the cloud drive it caches. So let's say I have a 30 GB drive and around 6-8 GB get allocated to the operating system and software, I could set a cache cap at 20 GB and not worry about the system running out of disk space.

**Comment by seuffert** [comment]
@felixbuenemann there is already an issue #1830 for this

**Comment by ncw** [comment]
I think the new cache backend serves this purpose, so I'm going to close this issue
